File: benchmark/parallel_array_write.cc
//    (c) University of Western Australia
//    International Centre of Radio Astronomy Research
//    M468, 35 Stirling Hwy
//    Crawley, Perth WA 6009
//    Australia
//
//    Shanghai Astronomical Observatory, Chinese Academy of Sciences
//    80 Nandan Road, Shanghai 200030, China
//
//    This library is free software: you can redistribute it and/or
//    modify it under the terms of the GNU General Public License as published
//    by the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    This library is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License along
//    with this library. If not, see <http://www.gnu.org/licenses/>.
//
//    Any bugs, questions, concerns and/or suggestions please email to
//    lbq@shao.ac.cn, jason.wang@icrar.org



// ################################################################################
// This code creates separate casa table files for each MPI process.
// In order to fool the casa table system as if different processes are
// dealing with independent casa tables so that one process won't lock
// the table and/or prevent others from putting data in, only the master MPI process
// creates casa table files in the intended place, while slave processes
// create table files with exactly the same contents in /tmp, which are supposed
// to be cleaned up once the job is finished.
//
// However, the AdiosStMan objects associated with these casa tables are actually all
// writing data into a single ADIOS container, which is in the same directory where
// the master MPI process writes casa table files. Therefore, after job is finished,
// these table files together with ADIOS files will contain all information that is
// necessary to reproduce the casa table.



#include <casacore/tables/Tables/SetupNewTab.h>
#include <casacore/tables/Tables/ScaColDesc.h>
#include <casacore/tables/Tables/ArrColDesc.h>
#include <casacore/tables/Tables/ArrayColumn.h>
#include <casacore/casa/namespace.h>
#include "../AdiosStMan.h"
#include "tictak.h"


int NrRows;
IPosition data_pos;
string filename;

int mpiRank, mpiSize;
Array<Float> data_arr;

int writeBufRows;

int rows_per_process;

void write_table(){
     
     rows_per_process = NrRows/mpiRank;

//    AdiosStMan stman("MPI_AGGREGATE", "num_aggregators=24, num_ost=24", 30000, 2000);
     AdiosStMan stman("POSIX", "", writeBufRows, rows_per_process);

    // define a table description & add a scalar column and an array column
    TableDesc td("", "1", TableDesc::Scratch);
    td.addColumn (ArrayColumnDesc<Float>("data", data_pos, ColumnDesc::Direct));

    // create a table instance, bind it to the storage manager & allocate rows
    SetupNewTable newtab(filename, td, Table::New);
    newtab.bindAll(stman);
    Table casa_table(newtab, NrRows);

    // define column objects and link them to the table
    ArrayColumn<Float> data_col(casa_table, "data");

    // each mpi rank writes a subset of the data
    for (uInt i=0; i<rows_per_process; i++) {
        data_col.put (mpiRank * rows_per_process + i, data_arr);
    }
}


int main (int argc, char **argv){

    MPI_Init(0,0);
    MPI_Comm_rank(MPI_COMM_WORLD, &mpiRank);
    MPI_Comm_size(MPI_COMM_WORLD, &mpiSize);

    if(argc < 6){
        cout << "./parallel_array_write (int)nrRows (int)arrayX (int)arrayY (string)filename (int)writeBufNrRows" << endl;
        exit(1);
    }

    NrRows = atoi(argv[1]);
    filename = argv[4];
    writeBufRows = atoi(argv[5]);

    data_pos = IPosition(2, atoi(argv[2]), atoi(argv[3]));
    data_arr = Array<Float>(data_pos);

    if(NrRows<mpiSize){
        exit(-1);
    }
    // generate filenames for slave processes
    // these files are not used later on, so just put them
    // into /tmp and clean them up when job is finished
    if(mpiRank>0){
        stringstream filename_s;
        filename_s << "/scratch/pawsey0129/AdiosStMan/tmp1/rank" << mpiRank << ".casa";
        filename = filename_s.str();
    }

    // put some data into the data array
    indgen (data_arr);

    MPI_Barrier(MPI_COMM_WORLD);
    tictak_add((char*)filename.c_str(),0);
    write_table();
    MPI_Barrier(MPI_COMM_WORLD);
    tictak_add((char*)"end",0);

    if(mpiRank == 0){

        float Seconds = tictak_total(0,0);
        uint64_t CellSize = atoi(argv[2])*atoi(argv[3])*sizeof(float);
        uint64_t TableSize = CellSize * NrRows;
        int Mps = TableSize / Seconds / 1000000;

        cout << "MB/s," << Mps;
        cout << ",Seconds," << Seconds;
        cout << ",TableSize(Byte)," << TableSize;
        cout << ",NrRows," << NrRows;
        cout << ",CellSize(Byte)," << CellSize;
        cout << ",MpiSize," << mpiSize;
        cout << ",Xlength," << atoi(argv[2]);
        cout << ",Ylength," << atoi(argv[3]);
        cout << endl;
    }

    MPI_Finalize();
    return 0;
}

File: benchmark/parallel_array_write.py
#!/usr/bin/python
#    parallel_array_write.py: Python script for benchmarking storage managers,
#	 collecting & plotting results
#
#    (c) University of Western Australia
#    International Centre of Radio Astronomy Research
#    M468, 35 Stirling Hwy
#    Crawley, Perth WA 6009
#    Australia
#
#    This library is free software: you can redistribute it and/or
#    modify it under the terms of the GNU General Public License as published
#    by the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This library is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License along
#    with this library. If not, see <http://www.gnu.org/licenses/>.
#
#    Any bugs, questions, concerns and/or suggestions please email to
#    jason.wang@icrar.org



import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import commands
import numpy
import os

ppn = 24
array_size = 16

node_max = 0
node_min = 100

row_max = 1
row_min = 10000

mbps_max = 1
mbps_min = 10000

def check_boundaries(linedic):

	global row_max, row_min, mbps_max, mbps_min, node_max, node_min


	if linedic['NrRows'] > row_max:
		row_max = int(linedic['NrRows'])

	if linedic['NrRows'] < row_min:
		row_min = int(linedic['NrRows'])

	if linedic['Mps'] > mbps_max:
		mbps_max = linedic['Mps']

	if linedic['Mps'] < mbps_min:
		mbps_min = linedic['Mps']

	if linedic['MpiSize']/ppn > node_max:
		node_max = int(linedic['MpiSize'])/ppn

	if linedic['MpiSize']/ppn < node_max:
		node_min = int(linedic['MpiSize'])/ppn


def add_point(linedic):


	#######
	plt.figure(1)
	color = (linedic['NrRows']-row_min)/(row_max-row_min)
	plt.scatter(linedic['MpiSize']/ppn, linedic['Mps'], c=(color, 1 - color , 0), s=5, edgecolors='none')

	#######
	plt.figure(2)
	color = (linedic['MpiSize']/ppn - node_min) / (node_max - node_min)
	if color > 1:
		color = 1
	plt.scatter(linedic['NrRows'], linedic['Mps'], c=(color, 1 - color , 0), s=5, edgecolors='none')


def save_fig():
	print node_max
	print node_min

	#######
	plt.figure(1)
	for i in range(row_min, row_max+1, 1000):
		colour = (float)(i-row_min)/(row_max-row_min)
		plt.scatter([],[], c=(colour, 1 - colour , 0), s=20, edgecolors='none', label=str("{0} Rows".format(i)))
	plt.legend(loc = 4)
	plt.grid()
	plt.xlim([node_min - 1, node_max + 10])
	plt.ylim([mbps_min - 100, mbps_max + 100])
	plt.xlabel('Number of Nodes ({0} processes per node)'.format(ppn))
	plt.ylabel('Total throughput in MB/s (cache saturated)')
	plt.figtext(.10, .94, "Parallel AdiosStMan (v0.3.6) testing, iVEC Fornax, Lustre Stripe Size = 8, \nSingle column with direct arrays, Array size = {0}MB, Table size = {0}MB * Rows".format(array_size))
	plt.savefig('fig1.eps')

	#######
	plt.figure(2)
	for i in range(10, node_max, 10):
		color = (float)(i)/node_max
		plt.scatter([],[], c=(color, 1 - color , 0), s=20, edgecolors='none', label=str("{0} Nodes".format(i)))
	plt.legend()
	plt.grid()
	plt.xlim([row_min - 100, row_max + 100])
	plt.ylim([mbps_min - 100, mbps_max + 100])
	plt.xlabel('Number of Rows')
	plt.ylabel('Total throughput in MB/s (cache saturated)')
	plt.figtext(.10, .94, "Parallel AdiosStMan (v0.3.6) testing, iVEC Fornax, Lustre Stripe Size = 8, \nSingle column with direct arrays, Array size = {0}MB, Table size = {0}MB * Rows".format(array_size))
	plt.savefig('fig2.eps')

#######  main

logfile = open('log','r')

for line in logfile:

	linedic = {}

	try:
		linelist = line.strip().split(',')
		for i in range(len(linelist)/2):
			linedic[linelist[i*2]] = float(linelist[i*2+1])
		check_boundaries(linedic)
	except:
		continue


logfile = open('log','r')

for line in logfile:

	linedic = {}

	try:
		linelist = line.strip().split(',')
		for i in range(len(linelist)/2):
			linedic[linelist[i*2]] = float(linelist[i*2+1])
		add_point(linedic)
	except:
		continue

save_fig()








